Allow setting open_threads and compression_level

rhpvorderman opened the pull request:
A very minor addition to the open function, but making it much more convenient. Compatibility with cutadapt is kept.

This fixes #90 . 

Discussion points:
- compresslevel vs compression_level. I think that compresslevel is not a very good abbreviation, on the other hand it is used consistently throughout CPython (and also xopen). Since dnaio is not a compression library I think the non-abbreviated version is better here.
- open_threads 0 by default. As someone who regularly works on a compute cluster I am regularly annoyed by applications that by default use all threads or a lot of them. One reason is that when I start a simple script/program on the login node this shouldn't immediately drain all the CPUs if I haven't told it so. This makes the login node unusable, so in my mind the only sane default is 1 thread, unless "parallel" is in the name of the application. (Pigz thus gets a free pass for wanting to use all threads). The other reason is that extra threads come with extra overhead. Since I usually run several samples at the same time I rather use less total compute time as I can run everything in parallel anyway to drive the wall clock time down. 
 
My opinions on the above are not absolute. As long as the open_threads parameter goes in I can choose a default for my own applications anyway. So I wonder what you think. As some extra information here are the benchmarks:

Running with open_threads 0. Total compute time 2.571 + 0.054 = 2.625
```
Benchmark 1: python dnaio_read.py ~/test/ERR039854.fastq.gz
  Time (mean ± σ):      2.625 s ±  0.085 s    [User: 2.571 s, System: 0.054 s]
  Range (min … max):    2.516 s …  2.748 s    10 runs
``` 

Running with open_threads=1. Decompress process is igzip version 2.30 (the fastest option). Total compute time 2.391 + 0.454 = 2.845
```
Benchmark 1: python dnaio_read.py ~/test/ERR039854.fastq.gz
  Time (mean ± σ):      1.560 s ±  0.035 s    [User: 2.391 s, System: 0.454 s]
  Range (min … max):    1.521 s …  1.626 s    10 runs
 ```

For comparison: reading the uncompressed file. The system time goes up a bit compared to 0 threaded gzip open because the filesystem is queried less on the smaller gzip file.
```
Benchmark 1: python dnaio_read.py ~/test/ERR039854.fastq
  Time (mean ± σ):      1.131 s ±  0.024 s    [User: 1.000 s, System: 0.131 s]
  Range (min … max):    1.107 s …  1.184 s    10 runs
```
 
My conclusion is that there is quite a penalty to be payed for using an external proess in terms of total compute time. This wasn't the case when I had just made python-isal, but by this point it is so optimized that it is much better than piping trough igzip.

marcelm commented:
I’m inclined to agree with the `open_threads=0` default, but give me a couple of days to think about this, and also the other change.

marcelm commented:
Can you clarify: The first benchmark result is for *not* for `open_threads=0`, I assume?

rhpvorderman commented:
Whoops an oversight in my text I see. I have two 0's. Now it correctly states that the first test is open_threads=0 and the second test is open_threads=1. 
As you can see, open_threads=0 leads to a longer wall-clock time, but a shorter compute time. open_threads=1 has a shorter wall-clock time, but pays for the overhead of the pipe and the threads with extra compute time.

marcelm reviewed (approved):
So the reason I’ve not gotten back to this for so long is that I thought that this would also change the default compression level to 1. However, it appears that this is already implicitly the case when using any xopen version >=1.0.0 where we switched to using an external `igzip` process by default because `igzip` has default compression level 1. Sorry for the delay.

Can you also add a changelog entry? Then we can merge this and release 0.10.0 as you suggested.

rhpvorderman commented:
> So the reason I’ve not gotten back to this for so long is that I thought that this would also change the default compression level to 1. However, it appears that this is already implicitly the case when using any xopen version >=1.0.0 where we switched to using an external igzip process by default because igzip has default compression level 1. Sorry for the delay.

No problem. I think this change will do much good now that the compression is transparantly and explicitly set to 1 always. That will lead to less unpredictable behaviour. Same goes for setting compression threads to 0.

Updated the changelog entry also with the changes from the multiple API. As well as fixing a typing error that mypy caught.

marcelm commented:
> I think this change will do much good now that the compression is transparantly and explicitly set to 1 always. That will lead to less unpredictable behaviour. Same goes for setting compression threads to 0.

Yes, it was a bit fuzzy before; this is an improvement.

> Updated the changelog entry also with the changes from the multiple API. As well as fixing a typing error that mypy caught.

Thanks.

> EDIT: Also did a double check on the latest cutadapt main to see if this API change does not break cutadapt. No test failures, so it should be fine. (Famous last words...)

I’ll merge this and then also run a check myself, just to be sure.
